java/arcs/core/util/Scheduler.kt

```diff
@@ -0,0 +1,240 @@
+    val loops = atomic(0)
+    private var processingJob: Job? = null
+    private val scope = CoroutineScope(context)
+    private val isIdle = atomic(true)
+    private val isPaused = atomic(false)
+    private val agenda = atomic(Agenda())
+
+    /** Schedule a single [Task] to be run as part of the next agenda. */
+    fun schedule(task: Task) {
+        agenda.update { it.addTask(task) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Schedule some [Task]s to be run as part of the next agenda. */
+    fun schedule(tasks: Iterable<Task>) {
+        agenda.update { it.addTasks(tasks) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Pause evaluation of the agenda. */
+    fun pause() {
+        isPaused.value = true
+    }
+
+    /** Resume evaluation of scheduled [Task]s in the agenda. */
+    fun resume() {
+        if (!isPaused.compareAndSet(expect = true, update = false)) return
+        // If we were paused, we can re-start.
+        scope.startProcessingJob()
+    }
+
+    /** Wait for the [Scheduler] to become idle. */
+    suspend fun waitForIdle() {
+        processingJob?.join()
+    }
+
+    private fun CoroutineScope.startProcessingJob() {
+        isIdle.value = false
+        processingJob = launch {
+            launches.incrementAndGet()
+
+            var shallContinue = true
+            while (shallContinue) {
+                val startTime = time.currentTimeMillis
+                shallContinue = withTimeout(agendaProcessingTimeout) {
+                    suspendCancellableCoroutine<Boolean> {
+                        it.resume(process()) { throwable ->
+                            if (throwable is TimeoutCancellationException) {
+                                log.error(throwable) { "Scheduled tasks timed out." }
+                            }
+                        }
+                    }
+                }
+                val elapsed = time.currentTimeMillis - startTime
+
+                if (shallContinue) {
+                    loops.incrementAndGet()
+
+                    // If the operation was super fast, let's delay a little bit to allow the
+                    // thread to chill out.
+                    val delayLength = scheduleRateHz.hzToMillisPerIteration() - elapsed
+                    if (delayLength > 0) delay(delayLength)
+                }
+            }
+        }.also { it.invokeOnCompletion { isIdle.value = true } }
+    }
+
+    private fun process(): Boolean {
+        if (isPaused.value) return false
+
+        val agenda = agenda.getAndSet(Agenda())
+        if (agenda.isEmpty()) return false
+
+        log.debug { "Processing agenda: $agenda" }
+
+        // Process agenda
+        agenda.processors.forEach { it() }
+        agenda.listenersByNamespace.listeners.asSequence()
+            .flatMap { (_, listenersByName) -> listenersByName.listeners.values.asSequence() }
+            .flatten()
+            .forEach { it() }
```

Comment:
Should this logic be inside the `Agenda` class? Are there other ways we might want to process an agenda?

Reply:
good point, I should probably make `Agenda` iterable.